### PR TITLE
🌱 add MachineDeployment scale test

### PR DESCRIPTION
**Pull request by jackfrancis**
**What this PR does / why we need it**:

This PR adds a new E2E test to validate MachineDeployment scale scenarios.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: test/framework/machinedeployment_helpers.go [hunk 1]
```diff
@@ -306,7 +306,7 @@ func ScaleAndWaitMachineDeployment(ctx context.Context, input ScaleAndWaitMachin
 	Expect(input.ClusterProxy).ToNot(BeNil(), "Invalid argument. input.ClusterProxy can't be nil when calling ScaleAndWaitMachineDeployment")
 	Expect(input.Cluster).ToNot(BeNil(), "Invalid argument. input.Cluster can't be nil when calling ScaleAndWaitMachineDeployment")
 
-	log.Logf("Scaling machine deployment %s/%s from %v to %v replicas", input.MachineDeployment.Namespace, input.MachineDeployment.Name, input.MachineDeployment.Spec.Replicas, input.Replicas)
+	log.Logf("Scaling machine deployment %s/%s from %d to %d replicas", input.MachineDeployment.Namespace, input.MachineDeployment.Name, *input.MachineDeployment.Spec.Replicas, input.Replicas)
```
Comment (jackfrancis) [inline test/framework/machinedeployment_helpers.go:309]: I noticed that this log output was returning a memory reference, so fixing this while I'm here!

File: test/e2e/md_scale_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+package e2e
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("When testing MachineDeployment scale out", func() {
```
Comment (enxebre) [inline test/e2e/md_scale_test.go:27]: nit: scale in/out?

**Comment by enxebre** [comment]
I'm not familiar with the e2e costs and duration, If there were concerns to keep those shrunk one could argue that this is implicitly tested by [cluster upgrade and drain timeout](https://github.com/kubernetes-sigs/cluster-api/search?q=ScaleAndWaitMachineDeployment&type=). Otherwise this is good to have and lgtm.

File: test/framework/machinedeployment_helpers.go (hunk 1 above)
Comment (JoelSpeed) [reply test/framework/machinedeployment_helpers.go:309]: What happens if the replicas are nil here? Is this a possibility?

Comment (jackfrancis) [reply test/framework/machinedeployment_helpers.go:309]: @mboersma @CecileRobertMichon should we protect against nil Replicas value here? (wouldn't that be a catastrophic error that needs handling somewhere earlier in the flow, if it's not already?)

Comment (CecileRobertMichon) [reply test/framework/machinedeployment_helpers.go:309]: maybe we can add an Expect above to make sure it's not nil?

**Comment by CecileRobertMichon** [comment]
In general i'm +1 to this change since it will be better to have a clear signal on scale up _and_ scale down than having to rely on the scale in the upgrade test.

lgtm besides nits above

File: test/e2e/md_scale.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+				KubernetesVersion:        input.E2EConfig.GetVariable(KubernetesVersion),
+				ControlPlaneMachineCount: pointer.Int64Ptr(1),
+				WorkerMachineCount:       pointer.Int64Ptr(1),
+			},
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		}, clusterResources)
+
+		Expect(clusterResources.MachineDeployments[0].Spec.Replicas).ToNot(BeNil())
```
Comment (jackfrancis) [inline test/e2e/md_scale.go:91]: @JoelSpeed @CecileRobertMichon I added these lines, I think this is the right place to deal with ensuring the hygiene of the replicas property. If that property is nil, we want to know about it here.

File: test/e2e/node_drain_timeout.go [hunk 4]
```diff
@@ -95,6 +95,8 @@ func NodeDrainTimeoutSpec(ctx context.Context, inputGetter func() NodeDrainTimeo
 		}, clusterResources)
 		cluster := clusterResources.Cluster
 		controlplane = clusterResources.ControlPlane
+		Expect(clusterResources.MachineDeployments[0].Spec.Replicas).ToNot(BeNil())
+		Expect(*clusterResources.MachineDeployments[0].Spec.Replicas).To(Equal(1))
```
Comment (jackfrancis) [inline test/e2e/node_drain_timeout.go:99]: @JoelSpeed @CecileRobertMichon more Replicas checking

File: test/framework/machinedeployment_helpers.go [hunk 5]
```diff
@@ -70,6 +70,7 @@ func GetMachineDeploymentsByCluster(ctx context.Context, input GetMachineDeploym
 
 	deployments := make([]*clusterv1.MachineDeployment, len(deploymentList.Items))
 	for i := range deploymentList.Items {
+		Expect(deploymentList.Items[i].Spec.Replicas).ToNot(BeNil())
```
Comment (jackfrancis) [inline test/framework/machinedeployment_helpers.go:73]: @JoelSpeed @CecileRobertMichon more Replicas non-nil hygiene

File: test/e2e/md_scale.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	"fmt"
+	"os"
+	"path/filepath"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+
```
Comment (srm09) [inline test/e2e/md_scale.go:30]: nit: remove empty line

File: test/e2e/md_scale.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+		Expect(*clusterResources.MachineDeployments[0].Spec.Replicas).To(Equal(1))
+
+		By("Scaling the MachineDeployment out to 3")
+		framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+			ClusterProxy:              input.BootstrapClusterProxy,
+			Cluster:                   clusterResources.Cluster,
+			MachineDeployment:         clusterResources.MachineDeployments[0],
+			Replicas:                  3,
+			WaitForMachineDeployments: input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
```
Comment (srm09) [inline test/e2e/md_scale.go:101]: Would adding an assertion for the new number of replicas make sense here?

File: test/e2e/md_scale.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+		By("Scaling the MachineDeployment down to 1")
+		framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+			ClusterProxy:              input.BootstrapClusterProxy,
+			Cluster:                   clusterResources.Cluster,
+			MachineDeployment:         clusterResources.MachineDeployments[0],
+			Replicas:                  1,
+			WaitForMachineDeployments: input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
```
Comment (srm09) [inline test/e2e/md_scale.go:109]: Same as above

File: test/e2e/md_scale.go (hunk 7 above)
Comment (jackfrancis) [reply test/e2e/md_scale.go:101]: `ScaleAndWaitMachineDeployment` does that for us in this case. The earlier assertion is to get a validated baseline before going through the scaling flow(s).

File: test/e2e/md_scale.go (hunk 8 above)
Comment (jackfrancis) [reply test/e2e/md_scale.go:109]: ditto

File: test/e2e/md_scale.go (hunk 7 above)
Comment (srm09) [reply test/e2e/md_scale.go:100]: Gotcha, I saw the implicit assertion now. During my earlier pass, I was looking for an explicit assertion in the `ScaleAndWaitMD` and hence the comment. Make sense.

File: test/e2e/md_scale.go (hunk 3 above)
Comment (jackfrancis) [reply test/e2e/md_scale.go]: in fact after further investigation the strict nil checks against Replicas aren't required as long as we use the ginkgo `Equal(` method, as that method includes nil checks, and if nil returns a specific error about that.

**Comment by jackfrancis** [comment]
@fabriziopandini this is ready for final review

File: test/e2e/md_scale.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+// MachineDeploymentScaleSpecInput is the input for MachineDeploymentScaleSpec.
+type MachineDeploymentScaleSpecInput struct {
```
Comment (fabriziopandini) [inline test/e2e/md_scale.go:36]: future work: expose the number of worker nodes we scale up to, so people reusing things Spec on providers might decide to have more (or less) workers according to their needs. As soon as we have HollowMachines in CAPD I plan to increase this for CAPI

**Review by fabriziopandini — COMMENTED** [review]
@jackfrancis Thanks for adding a new test to the CAPI E2E test suite! 
/lgtm

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**